### DISCUSSION TITLE
Merge Keys and Alias:

### DISCUSSION THREAD

**Discussion by BariHub** [aaubry/YamlDotNet#789] [Q&A]
I am currently using C# on Visual Studio 2022 and have been parsing some YAML files and would like to know how do you create merge keys and alias' nodes. For example, I have constructed a YamlStream that contain the following information:

      YamlMappingNode wRoot = new YamlMappingNode();
      YamlSequenceNode wWorkPositions = new YamlSequenceNode();
      YamlMappingNode wImmersiveDeveloper = new YamlMappingNode();
      wImmersiveDeveloper.Add(new YamlScalarNode("Description"), new YamlScalarNode("Developer"));
      wImmersiveDeveloper.Add(new YamlScalarNode("Level"), new YamlScalarNode("6"));
      wImmersiveDeveloper.Add(new YamlScalarNode("Location"), new YamlScalarNode("In the Office"));
      wImmersiveDeveloper.Anchor = "DEVELOPER";
      YamlMappingNode wJobDesc = new YamlMappingNode();
      wJobDesc.Add(new YamlScalarNode("JobDescription"), wImmersiveDeveloper);
      wWorkPoisition.Add(wJobDesc);

      YamlMappingNode wSpecialist = new YamlMappingNode();
      wImmersiveDeveloper.Add(new YamlScalarNode("Description"), new YamlScalarNode("Specialist"));
      wImmersiveDeveloper.Add(new YamlScalarNode("Level"), new YamlScalarNode("9"));
      wImmersiveDeveloper.Add(new YamlScalarNode("Location"), new YamlScalarNode("In the Lab"));
      wImmersiveDeveloper.Anchor = "IS";
      YamlMappingNode wJobDesc = new YamlMappingNode();
      wJobDesc.Add(new YamlScalarNode("JobDescription"), wSpecialist);
      wWorkPoisition.Add(wJobDesc);

      YamlMappingNode wGroupLeader = new YamlMappingNode();
      wImmersiveDeveloper.Add(new YamlScalarNode("Description"), new YamlScalarNode("Group Leader"));
      wImmersiveDeveloper.Add(new YamlScalarNode("Level"), new YamlScalarNode("10"));
      wImmersiveDeveloper.Add(new YamlScalarNode("Location"), new YamlScalarNode("In the Conference Room"));
      wImmersiveDeveloper.Anchor = "GL";
      YamlMappingNode wJobDesc = new YamlMappingNode();
      wJobDesc.Add(new YamlScalarNode("JobDescription"), wGroupLeader);
      wWorkPoisition.Add(wJobDesc);

      wRoot.Add(new YamlScalarNode("Jobs"), wWorkPositions);

      YamlMappingNode wEmployeeInfo = new YamlMappingNode();
      wEmployeeInfo.Add(new YamlScalarNode("first name"), new YamlScalarNode("dwayne"));
      wEmployeeInfo.Add(new YamlScalarNode("last name"), new YamlScalarNode("johnson"));

      YamlSequenceNode wSkills = new YamlSequenceNode();
      wSkills.Add(new YamlScalarNode("C++"));
      wSkills.Add(new YamlScalarNode("C#"));
      wSkills.Add(new YamlScalarNode("Unreal"));
      wSkills.Add(new YamlScalarNode("Javascript"));
      wSkills.Add(new YamlScalarNode("SQL"));

      wEmployeeInfo.Add(new YamlScalarNode("Skills"), wSkills);
      YamlMappingNode wAlias = new YamlMappingNode();
      wEmployeeInfo.Add(new YamlScalarNode("Current Position"), new YamlScalarNode("Unreal"));

      wRoot.Add(new YamlScalarNode("EmployeeInfo"), wEmployeeInfo);

      YamlDocument wDocument = new YamlDocument(wRoot);
      YamlStream wYamlStream = new YamlStream();
      wYamlStream.Add(wDocument);

      string wOutput;
      StringBuilder wStringBuilder = new StringBuilder();
      using (StringWriter = new StringWriter(wStringBuilder))
      {
        wYamlStream.Save(wWriter, false);
        wOutput = wStringBuilder.ToString();
      }

      Console.WriteLine(wOutput);`

If I were to output this stream to the console or a file, it will look like the following:

- JobDescription: &DEVELOPER
   Description: Developer
   Level: 6
   Location: In the Office
- JobDescription: &IS
   Description: Specialist
   Level: 9
   Location: In the Lab
- JobDescription: &GL
   Description: Group Leader
   Level: 10
   Location: In the Conference Room
EmployeeInfo:
   first name: dwayne
   last name: johnson
   skills:
   - C++
   - C#
   - unreal
   - javascript
   - SQL
   CurrentPosition: unreal

However, I would like to append the first job description to the bottom of the CurrentPosition Node using merge key/alias, like so:

- JobDescription: &DEVELOPER
   Description: Developer
   Level: 6
   Location: In the Office
- JobDescription: &IS
   Description: Specialist
   Level: 9
   Location: In the Lab
- JobDescription: &GL
   Description: Group Leader
   Level: 10
   Location: In the Conference Room
EmployeeInfo:
   first name: dwayne
   last name: johnson
   skills:
   - C++
   - C#
   - unreal
   - javascript
   - SQL
   CurrentPosition: unreal
   <<: *DEVELOPER

Is there a way to do this?

**Comment by Flatlid** [comment]
Hello @aaubry / @EdwardCooke 
I have come across a similar issue in the past using YamlDotNet. I had explored the sources and found `YamlAliasNode`, but that object is tagged as `internal`. In the end, I ended up not using Anchor/Alias for other reasons, so I didn't see a reason to open a discussion item, but seeing this post sure brings back memories. From my understanding, making `YamlAliasNode` this public would help resolve @BariHub 's problem. What do you think?

In the previous code snippet example, 
```
YamlMappingNode wAlias = new YamlMappingNode();
wEmployeeInfo.Add(new YamlScalarNode("Current Position"), new YamlScalarNode("Unreal"));
```
Would become:
```
YamlAliasNode wAlias = new YamlAliasNode(new AnchorName("DEVELOPER"));
wEmployeeInfo.Add(new YamlScalarNode("Current Position"), wAlias);
```

**Comment by EdwardCooke** [comment]
I wonder why that isn’t already a public class. I’ll take a look at it and see about making it public. There may be some underlying reasons it’s not.

**Comment by BariHub** [comment]
Thank you very much for replying, will keep a close eye on any updates!
